---
title: "Methylation analysis test with MethylKit"
output: html_notebook
author: Hernan Lorenzi
---

```{r load libraries}
library(methylKit)
library(cowplot)
library(genomation)

```


```{r Load data}
file.list.zKO <-list( 
                  "./data/GSE116489_RAW/GSM3240131_CpG_3L_4_methylKit.txt.gz",
                  "./data/GSE116489_RAW/GSM3240132_CpG_3L_5_methylKit.txt.gz",
                  "./data/GSE116489_RAW/GSM3240133_CpG_3L_6_methylKit.txt.gz",
                  "./data/GSE116489_RAW/GSM3240134_CpG_3L_7_methylKit.txt.gz",
                  "./data/GSE116489_RAW/GSM3240135_CpG_3L_9_methylKit.txt.gz")

# read the files to a methylRawList object: myobj
myobj <- methRead(location = file.list.zKO,
           sample.id=list("WT_1",   "WT_2",   "WT_3",   "zKO_1",  "zKO_2"),
           assembly="mm10",
           treatment=c(0,0,0,1,1),
           context="CpG",
           mincov = 20
           )

# read the gene BED file
gene.obj <- readTranscriptFeatures(location = "./data/UCSC_GENCODEvm23.tsv", up.flank = 1000, down.flank = 500, unique.prom = TRUE)

cpg.obj <- readFeatureFlank(location = "./data/UCSC_CpG_islands.tsv", remove.unusual = TRUE,
  flank = 0, clean = TRUE, feature.flank.name = NULL)

```

Since we read the methylation data now, we can check the basic stats about the methylation data such as coverage and percent methylation. We now have a methylRawList object which contains methylation information per sample. The following command prints out percent methylation statistics for second sample: “test2”
```{r}
getMethylationStats(myobj[[2]],plot=FALSE,both.strands=FALSE)
```


The following command plots the histogram for percent methylation distribution.The figure below is the histogram and numbers on bars denote what percentage of locations are contained in that bin. Typically, percent methylation histogram should have two peaks on both ends. In any given cell, any given base are either methylated or not. Therefore, looking at many cells should yield a similar pattern where we see lots of locations with high methylation and lots of locations with low methylation.

```{r}
dir.create(path = "./Plots", showWarnings = FALSE)
pdf(file = "./Plots/perc_meth_hist.pdf")
for (i in 1:length(myobj)){
  getMethylationStats(myobj[[i]],plot=TRUE,both.strands=FALSE)
}
dev.off()
```

We can also plot the read coverage per base information in a similar way, again numbers on bars denote what percentage of locations are contained in that bin. Experiments that are highly suffering from PCR duplication bias will have a secondary peak towards the right hand side of the histogram.

```{r}
pdf(file = "./Plots/read_cov_hist.pdf")
for (i in 1:length(myobj)){
  getCoverageStats(myobj[[i]],plot=TRUE,both.strands=FALSE)
}
dev.off()


```

It might be useful to filter samples based on coverage. Particularly, if our samples are suffering from PCR bias it would be useful to discard bases with very high read coverage. Furthermore, we would also like to discard bases that have low read coverage, a high enough read coverage will increase the power of the statistical tests. The code below filters a methylRawList and discards bases that have coverage below 10X and also discards the bases that have more than 99.9th percentile of coverage in each sample.

```{r}
filtered.myobj=filterByCoverage(myobj,lo.count=20,lo.perc=NULL,
                                      hi.count=NULL,hi.perc=99.9)
```

In order to do further analysis, we will need to get the bases covered in all samples. The following function will merge all samples to one object for base-pair locations that are covered in all samples. Setting destrand=TRUE (the default is FALSE) will merge reads on both strands of a CpG dinucleotide. This provides better coverage, but only advised when looking at CpG methylation (for CpH methylation this will cause wrong results in subsequent analyses). In addition, setting destrand=TRUE will only work when operating on base-pair resolution, otherwise setting this option TRUE will have no effect. The unite() function will return a methylBase object which will be our main object for all comparative analysis. The methylBase object contains methylation information for regions/bases that are covered in all samples.

```{r}
#meth=unite(filtered.myobj, destrand=TRUE, min.per.group = 2L, mc.cores = 16)
meth=unite(filtered.myobj, destrand=TRUE, mc.cores = 16)
meth
```

We can check the correlation between samples using getCorrelation. This function will either plot scatter plot and correlation coefficients or just print a correlation matrix.
```{r}
pdf("./Plots/sample_correlations.pdf")
  getCorrelation(meth,plot=FALSE)
dev.off()
```

We can cluster the samples based on the similarity of their methylation profiles. The following function will cluster the samples and draw a dendrogram.

```{r}
pdf("./Plots/sample_clustering_correl.pdf")
  clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)
dev.off()
clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)
```

We can also do a PCA analysis on our samples. The following function will plot a scree plot for importance of components.

```{r}
PCASamples(meth, screeplot=TRUE)
```

We can also plot PC1 and PC2 axis and a scatter plot of our samples on those axis which will reveal how they cluster.

```{r}
pdf("./Plots/pca_analysis.pdf")
  PCASamples(.Object = meth, comp = c(1, 2))
  PCASamples(.Object = meth, comp = c(1, 3))
  PCASamples(.Object = meth, comp = c(2, 3))
dev.off()

PCASamples(.Object = meth, comp = c(1, 2))
PCASamples(.Object = meth, comp = c(1, 3))
PCASamples(.Object = meth, comp = c(2, 3))
```
```{r}
myDiff=calculateDiffMeth(meth)
```
`
```{r get sign CpG}
myDiff.sign <- myDiff[myDiff$qvalue <= 0.05, ]

```


```{r annotate}

```